pkg/model/model.go

```diff
@@ -0,0 +1,282 @@
+	BlockDevice       *BlockDevice `json:"block_device"`
+	UUID              string       `json:"uuid"`
+	MountPoint        string       `json:"mount_point"`
+}
+
+// NewPartitionFromDelimited returns a Partition based upon
+// a delimited string value.
+func NewPartitionFromDelimited(delimited_string string, bd *BlockDevice) (p *Partition, err error) {
+	s_partition := strings.Split(delimited_string, ":")
+	p_pos, err := strconv.Atoi(s_partition[1])
+
+	if err != nil {
+		return
+	}
+
+	p, err = NewPartition(s_partition[0], uint(p_pos), s_partition[2], s_partition[3])
+	p.BlockDevice = bd
+
+	return
+}
+
+func NewPartition(name string, position uint, size string, ptype string) (p *Partition, err error) {
+	if position < 1 || position > 128 {
+		err = errors.New("Failed partitioning posititon: " + strconv.FormatUint(uint64(position), 10) + " A partition position must be >= 1 && <= 128")
+		return
+	}
+
+	p = &Partition{
+		Name:     name,
+		Position: position,
+		Size:     size,
+		Type:     ptype,
+	}
+
+	return
+}
+
+func NewBlockDevice(file string) (bd *BlockDevice, err error) {
+	bd = &BlockDevice{
+		File: file,
+	}
+
+	if !bd.Validate() {
+		err = errors.New("Block Device " + bd.File + " failed validation.")
+	}
+
+	return
+}
+
+// NewBlockDevices returns a slice of BlockDevice(s) for the supplied
+// slice of strings listing device files.
+func NewBlockDevices(s_devices ...string) (block_devices []*BlockDevice, err error) {
+	for _, dev := range s_devices {
+		bd, bd_err := NewBlockDevice(dev)
+		if err != nil {
+			return block_devices, bd_err
+		}
+
+		block_devices = append(block_devices, bd)
+	}
+
+	return
+}
+
+// GetDeviceFiles returns a slice of strings with all the device files
+// that make up the given RaidArray.
+// It returns a slice of strings and possibly an error.
+func (a RaidArray) GetDeviceFiles() (device_files []string, err error) {
+	for _, dev := range a.Devices {
+		device_files = append(device_files, dev.File)
+	}
+
+	return
+}
+
+// ValidateDevices validates that each block device is 'valid' by calling
+// Validate on each BlockDevice.
+// It returns false if any of the underlying calls to Validate() are false.
+func (a RaidArray) ValidateDevices() (valid bool) {
+	for _, bd := range a.Devices {
+		if !bd.Validate() {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Validate validates that the given block device is correct and accessible.
+// It returns an bool indicating pass/fail.
+func (b BlockDevice) Validate() bool {
+	resolved_path, err := filepath.EvalSymlinks(b.File)
+
+	if err != nil {
+		return false
+	}
+
+	fi, err := os.Stat(resolved_path) // Returns err if file is not accessible
+
+	if os.IsNotExist(err) {
+		return false
+	}
+
+	m := fi.Mode()
+
+	return m&os.ModeDevice != 0
+}
+
+// Format prepares a Partition on a given BlockDevice with a file system
+// It returns an error object, or nil depending on the results.
+func (p Partition) Format() (out string, err error) {
```

Comment:
recommend: pass `context.Context` to all methods that invoke OS commands, so that they can be canceled/executed with a timeout and we can add tracing with ease. 
This would then look like `Format(ctx context.Context)` and the context can be passed all the way down from the `cmd` package where this is invoked.